### PR TITLE
Array backend Part 3: Conversion and serialization

### PR DESCRIPTION
**Pull request by jangko**

### CODE REVIEW

File: stint/io.nim [hunk 1]
```diff
@@ -229,7 +231,9 @@ func readDecChar(c: range['0'..'9']): int {.inline.}=
   # specialization without branching for base <= 10.
   ord(c) - ord('0')
 
-func parse*[bits: static[int]](input: string, T: typedesc[StUint[bits]], radix: static[uint8] = 10): T =
+func parse*[bits: static[int]](input: string,
+                               T: typedesc[StUint[bits]],
+                               radix: static[uint8] = 10): T =
```
Comment (mratsim) [inline stint/io.nim:236]: This is the kind of function that shouldn't be made inline in my opinion.

There is a ``while`` loop inside.

File: stint/io.nim [hunk 2]
```diff
@@ -16,6 +16,10 @@ import
 
 from stew/byteutils import toHex
 
+# Helpers
+# --------------------------------------------------------
+{.push raises: [], inline, gcsafe.}
```
Comment (mratsim) [inline stint/io.nim:21]: I think ``inline`` should be selectively added here.

File: stint/io.nim [hunk 3]
```diff
@@ -104,87 +126,67 @@ func stuint*(a: StUint, bits: static[int]): StUint[bits] {.inline.} =
...
+    for i in 0 ..< result.len:
+      result[i] = a[i]
+  else:
+    for i in 0 ..< a.len:
+      result[i] = a[i]
+
+func stuint*(a: StInt, bits: static[int]): StUint[bits] {.inline.} =
+  ## signed int to unsigned int conversion
+  ## bigger to smaller bits conversion, the result is truncated
+  doAssert(a.isPositive, "Cannot convert negative number to unsigned int")
```
Comment (mratsim) [inline stint/io.nim:139]: ``OverflowDefect`` is supposed to handle Overflow and Underflow
or ``RangeDefect``

File: stint/io.nim [hunk 4]
```diff
@@ -104,87 +126,67 @@ func stuint*(a: StUint, bits: static[int]): StUint[bits] {.inline.} =
...
+  else:
+    result.impl = stuint(a.impl, bits)
+
+func stint*(a: StInt, bits: static[int]): StInt[bits] {.inline.} =
+  ## signed int to signed int conversion
+  ## will raise exception if input does not fit into destination
+  when a.bits < bits:
+    if a.isNegative:
+      result.impl = stuint(a.neg.impl, bits)
+      result.negate
```
Comment (mratsim) [inline stint/io.nim:155]: Negate can underflow

**Review by mratsim — COMMENTED** [review]
- inline should be added carefully on serialization proc to avoid size explosion
- consistent use of defects for Overflow and ranges

One thing is missing, handling of negating the smallest number.
for int8, the range is -128..127 so you can't negate -128, some of the serialization rely on negate and can likely be made to have the wrong result. THis needs to be tested.

File: stint/io.nim (hunk 1 above)
Comment (jangko) [reply stint/io.nim:236]: Yup, I didn't add push inline to that function.

File: stint/io.nim (hunk 2 above)
Comment (jangko) [reply stint/io.nim:21]: Fixed

File: stint/io.nim (hunk 3 above)
Comment (jangko) [reply stint/io.nim:139]: Fixed

File: stint/io.nim (hunk 4 above)
Comment (jangko) [reply stint/io.nim:155]: Right, special case for integer max/min should be added.

**Comment by mratsim** [comment]
I don't think there is a use-case for signed modular arithmetic besides just signed div and signed mod.

For addmod/mulmod/submod, the old stint only handled 0 ..< m with m positive modulus.

**Comment by jangko** [comment]
for completeness

**Pull request merged**